jupyter_kernel_mgmt/tests/test_kernelspec.py

```diff
@@ -24,177 +23,213 @@
-                                         user=False)
-
-    def test_remove_kernel_spec(self):
-        path = self.ksm.remove_kernel_spec('sample')
-        self.assertEqual(path, self.sample_kernel_dir)
-
-    def test_remove_kernel_spec_app(self):
-        p = Popen(
-            [sys.executable, '-m', 'jupyter_client.kernelspecapp', 'remove', 'sample', '-f'],
-            stdout=PIPE, stderr=STDOUT,
-            env=os.environ,
-        )
-        out, _ = p.communicate()
-        self.assertEqual(p.returncode, 0, out.decode('utf8', 'replace'))
-
-    def test_validate_kernel_name(self):
-        for good in [
-            'julia-0.4',
-            'ipython',
-            'R',
-            'python_3',
-            'Haskell-1-2-3',
-        ]:
-            assert kernelspec._is_valid_kernel_name(good)
-        
-        for bad in [
-            'has space',
-            u'ünicode',
-            '%percent',
-            'question?',
-        ]:
-            assert not kernelspec._is_valid_kernel_name(bad)
-
-    def test_provider_find_kernel_specs(self):
-        kernels = self.prov_ksm.find_kernel_specs()
-        self.assertEqual(kernels['prov_sample2'], self.prov_sample2_kernel_dir)
-        self.assertNotIn('sample', kernels)
-
-    def test_provider_get_kernel_spec(self):
-        ks = self.prov_ksm.get_kernel_spec('PROV_SAMPLE1')  # Case insensitive
-        self.assertEqual(ks.resource_dir, self.prov_sample1_kernel_dir)
-        self.assertEqual(ks.argv, sample_kernel_json['argv'])
-        self.assertEqual(ks.display_name, sample_kernel_json['display_name'])
-        self.assertEqual(ks.env, {})
-        self.assertEqual(ks.metadata, {})
-        with self.assertRaises(kernelspec.NoSuchKernel):
-            self.prov_ksm.get_kernel_spec('sample')
-
-    def test_provider_find_all_specs(self):
-        kernels = self.prov_ksm.get_all_specs()
-        self.assertEqual(len(kernels), 2)
-        self.assertEqual(kernels['prov_sample1']['resource_dir'], self.prov_sample1_kernel_dir)
-        self.assertIsNotNone(kernels['prov_sample1']['spec'])
+from .utils import install_sample_kernel, sample_kernel_json
+
+
+@pytest.fixture
+def sample_kernel_dir(setup_env):
+    kernel_dir = install_sample_kernel(pjoin(paths.jupyter_data_dir(), 'kernels'))
+    yield kernel_dir
+
+
+@pytest.fixture
+def prov_sample1_kernel_dir(setup_env):
+    kernel_dir = install_sample_kernel(
+        pjoin(paths.jupyter_data_dir(), 'kernels'), 'prov_sample1', 'prov_kernel.json')
+    yield kernel_dir
+
+
+@pytest.fixture
+def prov_sample2_kernel_dir(setup_env):
+
+    kernel_dir = install_sample_kernel(
+        pjoin(paths.jupyter_data_dir(), 'kernels'), 'prov_sample2', 'prov_kernel.json')
+    yield kernel_dir
+
+
+@pytest.fixture
+def ksm(setup_env):
+    spec_mgr = kernelspec.KernelSpecManager()
+    yield spec_mgr
+
+
+@pytest.fixture
+def prov_ksm(setup_env):
+    spec_mgr = kernelspec.KernelSpecManager(kernel_file='prov_kernel.json')
+    yield spec_mgr
```

Comment:
BTW, if you don't need to write any cleanup steps, your fixture can just return the value rather than yielding. Not important, but good to know.

Reply:
Right on - thank you.